### ISSUE TITLE
[Feature Request] Remove the White Flash Between Web Pages while Navigating in Dark Mode

### ISSUE THREAD

**Issue by kac487** [darkreader/darkreader#3811]
Hello,

I notice there is often a white flash that occurs when I am in dark mode and navigate between web pages. This is quite noticeable in certain settings like dark rooms. I expect this has something to do with some kind of latency between chrome loading the new page and then dark reader being applied. 

I really enjoy using this extension, but having a feature to keep the screen dark while navigating between web pages would make it that much better.   

Thanks

**Comment by Gusted** [comment]
Hi @kac487!

Do you have dark mode enabled in your system settings/preferences?

Regards,
Gusted

**Comment by kac487** [comment]
Yes i do. Changing the system dark theme doesn't seem to have much of an affect

**Comment by Gusted** [comment]
Do you have as well a dark theme installed for chrome?

For the rest we can't do anything, as #3533 was the best we can do in our abilities of a extension.

Regards,
Gusted

**Comment by ddnexus** [comment]
I had the same problem with Vivaldi (chrome based), which did actually solved the known "white flash of death" problem a few versions ago.

I discovered that Dark Reader reintroduces the problem with all the modes but Dynamic. 

Dynamic works well with no white flash, but Filter, Filter+ and Static do have the problem, although it is kind of random. The same pages from the same URLs may load with or without white flash. I couldn't find the exact reason (maybe caching, or connection speed, or who knows). 

What's certain is that Vivaldi does not have the problem without Dark Reader, and does not have it with Dark Reader in Dynamic mode.

Dynamic gives a better color/balance quality, but its problem is that it eats quite a few CPU cycles that the other mode don't. Even when you are not interacting with the page in any way and the browser is not even in the foreground, it still eats CPU. With the other filters that does not happen.

I guess with Dynamic there is some Javascript loop involved, and that could be acceptable while you are scrolling the page or interacting with it, but it should disappear on idle. Could that be improved?

**Comment by tmpm697** [comment]
@Gusted : does #3533 merged into master and being in latest release yet? As you said this might be the best effort to fix this `white` flash problem.

I still observe this issue with both static or dynamic.

Can we add a custom `dark` color page instead of white? Basically I want the same `dark` color of theme (`materia-gtk-theme` that I'm using)

**Comment by Gusted** [comment]
> static

That should be fixed by #3854 



> dynamic

If you open the console and type 
```Javascript
MatchMedia('(prefers-color-scheme: dark)').matches
```
What does it return?

**Comment by tmpm697** [comment]
I open `github.com` and open `console` tab in `inspect element` context menu, here result:
```
Uncaught ReferenceError: MatchMedia is not defined
    <anonymous> debugger eval code:1
debugger eval code:1:1
    <anonymous> debugger eval code:1
```

**Comment by Gusted** [comment]
Hmm seems to be a error, what about?
```javascript
window.matchMedia('(prefers-color-scheme: dark)').matches
```

**Comment by tmpm697** [comment]
result:
```
window.matchMedia('(prefers-color-scheme: dark)').matches
true
```

**Comment by Gusted** [comment]
Hmm, do you have a dark theme installed from the chrome webstore. As we are doing our best efforts to reduce the white flash however chrome has still flaws in the native code preventing to reduce the full white flash.

**Comment by tmpm697** [comment]
I use latest firefox/darkreader and dark theme from system. 

I don't use dark-theme for firefox from firefox store.

**Comment by Gusted** [comment]
Hmm, do you have shadowfox installed? They resolves a lot of white flash for me on Firefox.

**Comment by tmpm697** [comment]
I don't use shadowfox as it disables dynamic feature of darkreader.

**Comment by Gusted** [comment]
> 
> 
> I don't use shadowfox as it disables dynamic feature of darkreader.

What about just the White Flash part of shadowfox https://github.com/overdodactyl/ShadowFox/blob/master/css/userChrome-files/remove_white_flash.css to be included in userChrome.css

**Comment by YodaEmbedding** [comment]
> Dynamic works well with no white flash, but Filter, Filter+ and Static do have the problem, although it is kind of random. The same pages from the same URLs may load with or without white flash. I couldn't find the exact reason (maybe caching, or connection speed, or who knows).

Odd... I see the opposite: white flashes when using Dynamic due to slow loading on [somewhat heavy webpages](https://tex.stackexchange.com/questions/55210/not-element-of-in-latin-modern) whereas Filter[+] applies instantly.

Perhaps there might be something that can be borrowed from the `Dark Background Light Text` extension -- I never had white flashes with that extension, though its styling methods are of course a bit simplistic.

**Comment by Gusted** [comment]
> Odd... I see the opposite: white flashes when using Dynamic due to slow loading on somewhat heavy webpages whereas Filter[+] applies instantly.

You're sure you have the latest Dark Reader version, as the Filter[+] logic should technically apply later than the Dynamic logic.

> Perhaps there might be something that can be borrowed from the Dark Background Light Text extension -- I never had white flashes with that extension, though its styling methods are of course a bit simplistic. It seems to pre-style all the white areas with a dark gray while the page is still rendering.

Having 5 seconds looking at their logic and at our [fallback.ts](https://github.com/darkreader/darkreader/blob/master/src/inject/fallback.ts) that prevents the white flashes. They seems to have more overhead than the fallback.ts. 
So I'm not sure how you're getting dark flashes other than, if you open the Console in Dev Tools and enter 
```javscript
matchMedia('(prefers-color-scheme: dark)').matches;
```
What does it return?

**Comment by ddnexus** [comment]
@Gusted the `window.matchMedia('(prefers-color-scheme: dark)').matches` returns `flase  here. What is that supposed to mean?

**Comment by Gusted** [comment]
> 
> 
> @Gusted the `window.matchMedia('(prefers-color-scheme: dark)').matches` returns `false` here. What is that supposed to mean?

It checks if your system prefers a dark color scheme. E.g. in windows ![](https://i.imgur.com/l7W1N92.png) It will return `true`. This is one of the conditions requires to enable the fallback style to be injected. If this returns `false` it won't inject the fallback style and therefor a white flash.

**Comment by tmpm697** [comment]
@Gusted I have extract those lines in `userChrome.css` before you mention it. 

Even with low specs pc and low internet speed, I still expect/want this `white` page issue doesn't happen.

**Comment by ddnexus** [comment]
@Gusted Thank you for the explanation. I researched and the only way I could get the expression to return `true` has been starting Vivaldi with `--force-dark-mode`. 

I am still running the same version though, so the non-dynamic modes still have the white flash. I saw that you added #3854. How can I try that? 

Also, what about the constant use of extra CPU consumed by the dynamic mode?

**Comment by Gusted** [comment]
> Even with low specs pc and low internet speed, I still expect/want this white page issue doesn't happen.

Than I will say you need to go report a bug at [Mozilla's bugzilla](bugzilla.mozilla.org/) that you still have white flashes even after a userChrome.css change and Dark Reader what should 100% solve the issue for Geckoview based browser.

**Comment by Gusted** [comment]
> I am still running the same version though, so the non-dynamic modes still have the white flash. I saw that you added #3854. How can I try that?

[build.zip](https://github.com/darkreader/darkreader/files/5404070/build.zip)

> Also, what about the constant use of extra CPU consumed by the dynamic mode?

Dynamic means it analyze the styles of the site and generate a nice darkened style of it alongside a lot of other things.

**Comment by ddnexus** [comment]
> Dynamic means it analyze the styles of the site and generate a nice darkened style of it alongside a lot of other things.

I understand that. I guess it has to consume some CPU when it does that, however, once it is done, ]the dynamic mode should not consue CPU anymore (since it the "nice darkened style" should be already generated). Isn't that what is supposed to happen?

What are the "a lot of other things" that it is doing. Again, shouldn't that happen at the beginning and then consume virtually 0 CPU?

**Comment by ddnexus** [comment]
I installed the build that you provided, and it does not improve the random white flash for the non-dynamic modes.
I tried it with the `prefers-color-scheme` both `true` and `false` and it didn't change anything in the white flash for non-dynamic modes.
Dynamic still works well with no white flash at all. Still consuming CPU even on idle.

**Comment by Gusted** [comment]
> I understand that. I guess it has to consume some CPU when it does that, however, once it is done, ]the dynamic mode should not consue CPU anymore (since the "nice darkened style" should be already generated). Isn't that what is supposed to happen?

Their might be some edge cases, but in most cases just [this simply code](https://github.com/darkreader/darkreader/blob/master/src/inject/dynamic-theme/style-manager.ts#L340) is being executed in idle in order to watch style sheets that changes CSS.

> What are the "a lot of other things" that it is doing

Well, the dynamic mode isn't just checking style sheets and darkening them. To name some key functionality, analyze Background-images in order to possible invert them. Making sure a Optimized Observer is in the background to handle all DOM nodes that might possible contain new styles or removed/moved styles. 

> Again, shouldn't that happen at the beginning and then consume virtually 0 CPU?

It doesn't happen at the beginning as they are a step in the whole `dynamic mode step` and therefor only called when needed. After that it consume 0 CPU*.

> I installed the build that you provided, and it does not improve the random white flash for the non-dynamic modes.

That means I need another approach on the PR.

> I tried it with the prefers-color-scheme both true and false and it didn't change anything in the white flash for non-dynamic modes.

In order to kill the white flashes the `prefers-color-scheme` must return `true`.

> Still consuming CPU even on idle.

Their is [this simply code](https://github.com/darkreader/darkreader/blob/master/src/inject/dynamic-theme/style-manager.ts#L340) and if you might suspect their is some faulty code also consuming CPU on idle. Open up a issue for it.

**Comment by ddnexus** [comment]
> It doesn't happen at the beginning as they are a step in the whole dynamic mode step and therefor only called when needed. After that it consume 0 CPU*.

Good strategy. However it looks that - on most pages - it's "needed" constantly. The same page (gmail inbox for example) left idle for at least a 1 minute lowers to about 0 CPU with filter+, filter and static. With dynamic - in the same conditions - it never gets under 11-12%. 

For example, this very page is running at 9-10% too.

> Their is this simply code and if you might suspect their is some faulty code also consuming CPU on idle. Open up a issue for it.

Not sure there is any "faulty code", but it does consume CPU on idle. Maybe the pages are not so "idle" as we think they are, (so maybe a different strategy could avoid that?) but dynamic definitely consume a minimum of 10% more CPU than the other modes. That's when it is in the background. If you are interacting with the page it consume quite a lot more. That's the reason I would be OK using filter+ when the white flash will be fixed (or use dynamic if the CPU would be lower).

**Comment by ddnexus** [comment]
FYI: I just found a page that always white-flashes even with dynamic: https://searchengineland.com/google-no-longer-supports-relnext-prev-314319

Hoping that it ill be useful to you, but I think there is something wrong with the page itself, since it starts to display an empty dark background, then it turns white for a few seconds, and then finally the page arrive with the dark style applied.

**Comment by Gusted** [comment]
> Good strategy. However it looks that - on most pages - it's "needed" constantly. The same page (gmail inbox for example) left idle for at least a 1 minute lowers to about 0 CPU with filter+, filter and static. With dynamic - in the same conditions - it never gets under 11-12%.

I just checked with Gmail, after the loading icon is gone and the gmail logo is on the tab. Their is no code being runned other than the stylesheet checker as mentioned earlier. Maybe Gmail is still running code in the background *It's a google product after all with their shady Trackers and ads*. Could you try with uBlock origin? As when I stopped all code from gmail on Vivaldi when I saw a spike in the internal Task Manager. It was some obfuscated code loaded from a know Google ad CDN. with uBlock origin it was after around 10 seconds just 0% CPU usage constantly.

> Not sure there is any "faulty code", but it does consume CPU on idle. Maybe the pages are not so "idle" as we think they are, (so maybe a different strategy could avoid that?) but dynamic definitely consume a minimum of 10% more CPU than the other modes. That's when it is in the background. If you are interacting with the page it consume quite a lot more. That's the reason I would be OK using filter+ when the white flash will be fixed (or use dynamic if the CPU would be lower).

I'm not sure how you're measuring, but take a look at the internal task manager, as it might be the browser self that is executing code itself as well.
![](https://i.imgur.com/g5o8kQU.gif)

**Comment by Gusted** [comment]
> FYI: I just found a page that always white-flashes even with dynamic: https://searchengineland.com/google-no-longer-supports-relnext-prev-314319

Hoping that it ill be useful to you, but I think there is something wrong with the page itself, since it starts to display an empty dark background, then it turns white for a few seconds, and then finally the page arrive with the dark style applied.

Opening a separate issue for this as this is some other cause.

**Comment by Gusted** [comment]
#3884

**Comment by ddnexus** [comment]
I am not sure we are talking bout the same thing when we say "idle", so let me explain to avoid any misunderstanding.

I don't mean that gmail keeps consuming CPU when you switch the browser to another tab, especially to an empty tab as in your screenshot. In that case it will go down to 0 and come up just from time to time just for a fraction of a second. I am talking about when you don't interact with the page that is displayed in the tab, and you switch to another app.

In that case the CPU for the whole browser comes down almost to 0 CPU with filter+, filter and static, but it will stay at 10-12% minimum with dynamic. That does not change much (if at all) with gmail, or github or most other URLs. It still consumes more CPU even if the page is not doing anything.

BTW, the task manager shows the same as using `top` or using the Ubuntu System Monitor. The total CPU consumption is the same.

**Comment by RaidPX** [comment]
I agree that can get pretty annoying

**Comment by Gusted** [comment]
> I don't mean that gmail keeps consuming CPU when you switch the browser to another tab, especially to an empty tab as in your screenshot. In that case it will go down to 0 and come up just from time to time just for a fraction of a second. I am talking about when you don't interact with the page that is displayed in the tab, and you switch to another app.

So:
Opens tab e.g. gmail and don't touch it anymore
Without minimizing vivaldi or anything open another app e.g. Mail client.



> In that case the CPU for the whole browser comes down almost to 0 CPU with filter+, filter and static, but it will stay at 10-12% minimum with dynamic. That does not change much (if at all) with gmail, or github or most other URLs. It still consumes more CPU even if the page is not doing anything.

Hmm, I know certain sites which are just conflicting with Dark reader but those are mostly edge cases and not popular where their will still some CPU utilization. Could you check https://example.com which is just a simple site and check if you still have this 10-20% CPU Utilization.

**Comment by ddnexus** [comment]
example.com is a bit less... about 6-8% compared to Gmail that is 11-12%.
Dynamic involves more GPU process too. It's about twice as much compared to other modes. and seem to be proportional with the active tab (i.e. less for example.com and more for Gmail).

**Comment by Gusted** [comment]
> example.com is a bit less... about 6-8% compared to Gmail that is 11-12%.

Now it's for sure something shady going on as that site should even on the lowest spec PC's have 0 impact as it has little CSS.
Could you try and run a performance(ctrl + shift + i -> Performance tab) and press ctrl + shift + e after around 10 seconds if it didn't stop automatically stop it. And save the profile ![](https://i.imgur.com/GONjxKt.png)

**Comment by ddnexus** [comment]
Here are the 2 profiles for 10 sec for dynamic and for filter+. VERY different also in size!
[Profiles.zip](https://github.com/darkreader/darkreader/files/5422465/Profiles.zip)

**Comment by Gusted** [comment]
Moving the CPU utilization issue to #3905

**Comment by refael722** [comment]
Just suggest, maybe you can get around it, with a semi-transparent black screen, above the whole screen, like a pop up or something, and then after loading the site, it will disappear, so even if there is a flash, it will weaken it,

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
This is not a problem that is caused by Dark Reader. This problem should be reported at Mozilla's Bugzilla and Google's Bug Tracker.
